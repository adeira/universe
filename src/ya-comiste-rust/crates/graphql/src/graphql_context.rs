#[derive(Clone)]
pub struct Context {
    pub pool: arangodb::ConnectionPool,
}

impl juniper::Context for Context {}
