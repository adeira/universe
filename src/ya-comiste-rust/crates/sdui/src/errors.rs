/// Logic error (when something doesn't make any sense):
///
/// ```
/// use sdui::errors::ModelError;
/// assert_eq!(format!("{}", ModelError::LogicError("ups".to_string())), "Logic error: ups")
/// ```
pub enum ModelError {
    // TODO: naming (SDUIError?)
    DatabaseError(arangors::ClientError),
    LogicError(String),
}

// so we can use the `?` operator
impl From<arangors::ClientError> for ModelError {
    fn from(err: arangors::ClientError) -> ModelError {
        ModelError::DatabaseError(err)
    }
}

impl std::fmt::Display for ModelError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match *self {
            ModelError::DatabaseError(ref err) => write!(f, "Database error: {}", err.to_string()),
            ModelError::LogicError(ref err) => write!(f, "Logic error: {}", err),
        }
    }
}
