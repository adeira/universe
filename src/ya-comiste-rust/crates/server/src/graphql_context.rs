#[derive(Clone, Copy, Debug)]
pub struct Context;
impl juniper::Context for Context {}
