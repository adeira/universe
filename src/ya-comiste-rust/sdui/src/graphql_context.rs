use crate::model::component_content::ContentDataloaderType;
use crate::user::User;

#[derive(Clone)]
pub struct Context {
    pub pool: arangodb::ConnectionPool,
    pub content_dataloader: ContentDataloaderType,
    pub user: User,
}

impl juniper::Context for Context {}
