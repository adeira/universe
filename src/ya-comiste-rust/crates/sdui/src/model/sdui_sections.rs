use crate::errors::ModelError;
use crate::model::entrypoints::get_entrypoint;
use crate::sdui_component::SDUIComponent;
use crate::sdui_section::SDUISection;
use arangodb::connection;

/// Note: this function doesn't return `section.component` on purpose.
pub async fn get_all_sections_for_entrypoint_key(
    entrypoint_key: String,
) -> Result<Vec<SDUISection>, ModelError> {
    let conn = connection().await;
    let db = conn.db("ya-comiste").await.unwrap();

    let entrypoint = get_entrypoint(&entrypoint_key).await?;
    let aql = arangors::AqlQuery::builder()
        .query(
            "
            FOR section IN 1..1 OUTBOUND @entrypoint_id entrypoint_sections
            LET component = section.component
            SORT section.order ASC
            RETURN {
                id: section._id,
            }
            ",
        )
        .bind_var("entrypoint_id", entrypoint._id.to_string())
        .batch_size(1)
        .build();

    match db.aql_query::<SDUISection>(aql).await {
        Ok(r) => Ok(r),
        Err(e) => Err(ModelError::DatabaseError(e)),
    }
}

pub async fn get_section_component(
    section_id: String,
    supported: Vec<String>,
) -> Result<SDUIComponent, ModelError> {
    let conn = connection().await;
    let db = conn.db("ya-comiste").await.unwrap();

    // TODO: change the component ID to be unique (how to ðŸ¤”)
    let aql = arangors::AqlQuery::builder()
        .query(
            "
            LET section = DOCUMENT(@section_id)
            LET component = section.component
            FILTER component.typename IN @supported_typenames
            LIMIT 1
            RETURN {
                typename: component.typename,
                id: CONCAT(section._id, '~', component.typename),
            }
            ",
        )
        .bind_var("section_id", section_id)
        .bind_var("supported_typenames", supported)
        .batch_size(1)
        .build();

    match db.aql_query::<SDUIComponent>(aql).await {
        Ok(r) => match r.first() {
            Some(first) => Ok(first.to_owned()),
            None => Err(ModelError::LogicError(
                "no supported component available".to_string(),
            )),
        },
        Err(e) => Err(ModelError::DatabaseError(e)),
    }
}
