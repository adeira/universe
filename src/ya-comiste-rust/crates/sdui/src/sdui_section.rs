use crate::model::sdui_sections::get_section_components;
use crate::sdui_component::SDUIComponent;
use graphql::graphql_context::Context;
use juniper::{FieldError, FieldResult};

#[derive(Clone, Debug, serde::Deserialize)]
pub struct SDUISection {
    _id: String,
    _key: String,
    _rev: String,
    order: i16, // Â±32767
}

// impl SDUISection {
//     fn not_exposed_to_graphql(&self) -> &str {
//         &self.title
//     }
// }

#[juniper::graphql_object(context = Context)]
impl SDUISection {
    fn id(&self) -> juniper::ID {
        juniper::ID::new(&self._id)
    }

    async fn component(
        &self,
        supported: Vec<String>,
        context: &Context,
    ) -> FieldResult<Option<SDUIComponent>> {
        // We might get a component which is not supported by the client yet (defined by `supported`).
        let connection_pool = context.pool.to_owned();
        let components =
            get_section_components(connection_pool, self._id.to_string(), supported).await;
        match components {
            Ok(c) => Ok(match c.first() {
                Some(component) => Some(component.to_owned()),
                None => None, // TODO: log this missing supported component situation (?)
            }),
            Err(e) => Err(FieldError::from(e)),
        }
    }
}
