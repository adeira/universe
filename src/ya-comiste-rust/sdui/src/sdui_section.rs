use crate::graphql_context::Context;
use crate::model::sdui_sections::get_section_components;
use crate::sdui_component::SDUIComponent;
use juniper::{FieldError, FieldResult};

#[derive(Clone, Debug, serde::Deserialize)]
pub struct SDUISection {
    _id: String,
    _key: String,
    _rev: String,
    _from: String,
    _to: String,
    order: i16, // Â±32767
}

#[juniper::graphql_object(context = Context)]
impl SDUISection {
    fn id(&self) -> juniper::ID {
        juniper::ID::new(&self._id)
    }

    async fn component(
        &self,
        supported: Vec<String>,
        context: &Context,
    ) -> FieldResult<Option<SDUIComponent>> {
        // We might get a component which is not supported by the client yet (defined by `supported`).
        let connection_pool = context.pool.to_owned();
        let components = get_section_components(
            &context.user,
            &connection_pool,
            &self._from.to_string(),
            &self._id.to_string(),
            &supported,
        )
        .await;

        match components {
            Ok(c) => Ok(match c.first() {
                Some(component) => Some(component.to_owned()),
                None => {
                    log::warn!(
                        "No component found for section ID '{}' and supported selection: {}",
                        &self._id,
                        &supported.join(", ")
                    );
                    None
                }
            }),
            Err(e) => Err(FieldError::from(e)),
        }
    }
}
