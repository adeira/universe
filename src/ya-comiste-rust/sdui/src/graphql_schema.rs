use crate::graphql_context::Context;
use crate::model::sdui_sections::get_all_sections_for_entrypoint_key;
use crate::sdui_section::SDUISection;
use arangodb::errors::ModelError;
use auth::users::User;
use juniper::{EmptySubscription, FieldError, FieldResult, RootNode};

#[derive(Clone, Copy, Debug)]
pub struct Query;

#[derive(juniper::GraphQLObject)]
struct WhoamiPayload {
    id: Option<juniper::ID>,

    /// Human readable type should be used only for testing purposes. The format is not guaranteed
    /// and can change in the future completely.
    human_readable_type: Option<String>,
}

#[juniper::graphql_object(context = Context)]
impl Query {
    async fn mobile_entrypoint_sections(
        key: String,
        context: &Context,
    ) -> FieldResult<Vec<SDUISection>> {
        let connection_pool = context.pool.to_owned();
        match get_all_sections_for_entrypoint_key(&context.user, &connection_pool, &key).await {
            Ok(s) => Ok(s),
            // Err(e) => Err(FieldError::from(e)),
            Err(e) => match e {
                ModelError::DatabaseError(e) => Err(FieldError::from(e)), // TODO: hide and log these errors
                ModelError::LogicError(e) => Err(FieldError::from(e)),
                ModelError::SerdeError(e) => Err(FieldError::from(e)),
            },
        }
    }

    /// Returns information about the current user (can be authenticated or anonymous).
    async fn whoami(context: &Context) -> WhoamiPayload {
        match &context.user {
            User::AuthorizedUser(user) => WhoamiPayload {
                id: Some(juniper::ID::from(user.id())),
                human_readable_type: Some(String::from("authorized user")),
            },
            User::AnonymousUser(user) => WhoamiPayload {
                id: Some(juniper::ID::from(user.id())),
                human_readable_type: Some(String::from("anonymous user")),
            },
            User::UnauthorizedUser(user) => WhoamiPayload {
                id: Some(juniper::ID::from(user.id())),
                human_readable_type: Some(String::from("unauthorized (but not anonymous) user")),
            },
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Mutation;

#[derive(juniper::GraphQLObject)]
struct AuthorizeMobilePayload {
    success: bool,

    /// Session token should be send with every GraphQL request which requires auth.
    /// Returns `None` if the request was not successful.
    session_token: Option<String>,
}

#[derive(juniper::GraphQLObject)]
struct DeauthorizeMobilePayload {
    success: bool,
}

#[juniper::graphql_object(context = Context)]
impl Mutation {
    /// This function accepts Google ID token (after receiving it from Google Sign-In in a mobile
    /// device) and returns authorization payload. There is no concept of sign-in and sign-up
    /// because every user with a valid JWT ID token will be either authorized OR registered and
    /// authorized. Invalid tokens and disabled tokens will be rejected.
    async fn authorize_mobile(
        google_id_token: String,
        context: &Context,
    ) -> FieldResult<AuthorizeMobilePayload> {
        let connection_pool = context.pool.to_owned();
        let session_token = auth::authorize(&connection_pool, &google_id_token).await;
        match session_token {
            Ok(session_token) => Ok(AuthorizeMobilePayload {
                success: true,
                session_token: Some(session_token),
            }),
            Err(e) => {
                log::error!("{}", e);
                Ok(AuthorizeMobilePayload {
                    success: false,
                    session_token: None,
                    // TODO: return rejection reason from AuthError as well (?)
                })
            }
        }
    }

    /// The purpose of this `deauthorize` mutation is to remove the active sessions and efectivelly
    /// make the mobile application unsigned. Mobile applications should remove the session token
    /// once deauthorized.
    async fn deauthorize_mobile(
        session_token: String,
        context: &Context,
    ) -> DeauthorizeMobilePayload {
        let connection_pool = context.pool.to_owned();
        match auth::deauthorize(&connection_pool, &session_token).await {
            Ok(_) => DeauthorizeMobilePayload { success: true },
            Err(_) => DeauthorizeMobilePayload { success: false },
        }
    }
}

pub type Schema = RootNode<'static, Query, Mutation, EmptySubscription<Context>>;

pub fn create_graphql_schema() -> Schema {
    Schema::new(Query, Mutation, EmptySubscription::<Context>::new())
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    #[test]
    fn schema_snapshot() {
        // This test will make sure that the schema generated by server is ready to be used by
        // clients. How to do it better?

        let new_schema_path = "./../../ya-comiste-meta/schema.graphql.new";
        let saved_schema_path = "./../../ya-comiste-meta/schema.graphql";
        let saved_schema_snapshot =
            fs::read_to_string(saved_schema_path).expect("unable to read schema file");

        assert!(signedsource::is_signed(&saved_schema_snapshot));
        assert!(signedsource::is_valid_signature(&saved_schema_snapshot));

        let new_schema_snapshot = signedsource::sign_file(&format!(
            "# {}\n\n{}",
            signedsource::SIGNING_TOKEN,
            super::create_graphql_schema().as_schema_language()
        ));

        if saved_schema_snapshot != new_schema_snapshot {
            fs::write(new_schema_path, new_schema_snapshot)
                .expect("unable to write new schema file");
        }

        assert_eq!(
            Path::new(new_schema_path).exists(),
            false,
            "schema snapshot with *.new extension should not exist - please resolve it"
        );
    }
}
