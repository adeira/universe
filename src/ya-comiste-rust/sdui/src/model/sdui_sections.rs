use crate::errors::ModelError;
use crate::model::entrypoints::get_entrypoint;
use crate::sdui_component::SDUIComponent;
use crate::sdui_section::SDUISection;

pub async fn get_all_sections_for_entrypoint_key(
    pool: arangodb::ConnectionPool,
    entrypoint_key: String,
) -> Result<Vec<SDUISection>, ModelError> {
    let conn = pool.get().await.unwrap(); // TODO: DRY, no unwrap
    let db = conn.db("ya-comiste").await.unwrap(); // TODO: DRY, no unwrap

    let entrypoint = get_entrypoint(pool, &entrypoint_key).await?;
    let aql = arangors::AqlQuery::builder()
        .query(
            "
            FOR section IN 1..1 OUTBOUND @entrypoint_id entrypoint_sections
            SORT section.order ASC
            RETURN section
            ",
        )
        .bind_var("entrypoint_id", entrypoint._id.to_string())
        .build();

    match db.aql_query::<SDUISection>(aql).await {
        Ok(r) => Ok(r),
        Err(e) => Err(ModelError::DatabaseError(e)),
    }
}

pub async fn get_section_components(
    pool: arangodb::ConnectionPool,
    section_id: String,
    supported: Vec<String>,
) -> Result<Vec<SDUIComponent>, ModelError> {
    let conn = pool.get().await.unwrap(); // TODO: DRY, no unwrap
    let db = conn.db("ya-comiste").await.unwrap(); // TODO: DRY, no unwrap

    let aql = arangors::AqlQuery::builder()
        .query(
            "
            FOR component IN 1..1 OUTBOUND @section_id section_components
            FILTER component.typename IN @supported_typenames
            LIMIT 1
            RETURN {
                _serde_union_tag: component.typename,
                _serde_union_content: component,
            }
            ",
        )
        .bind_var("section_id", section_id)
        .bind_var("supported_typenames", supported)
        .build();

    let result = db.aql_query::<SDUIComponent>(aql).await;
    match result {
        Ok(r) => Ok(r),
        Err(e) => Err(ModelError::DatabaseError(e)),
    }
}
