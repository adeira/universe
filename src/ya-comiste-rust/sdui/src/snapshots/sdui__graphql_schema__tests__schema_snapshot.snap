---
source: sdui/src/graphql_schema.rs
expression: "super::create_graphql_schema().as_schema_language()"
---
type SDUIScrollViewHorizontalComponent {
  id: ID!
  title: String!
  cards: [SDUICardComponent!]!
}

type SDUIDescriptionComponent {
  id: ID!
  text: String!
}

type SDUIJumbotronComponent {
  id: ID!
  title: String!
}

type SDUICardComponent {
  id: ID!
  entrypointKey: String!
  title: String!
}

type Query {
  mobileEntrypointSections(key: String!): [SDUISection!]!
}

type SDUISection {
  id: ID!
  component(supported: [String!]!): SDUIComponent
}

union SDUIComponent = SDUICardComponent | SDUIDescriptionComponent | SDUIJumbotronComponent | SDUIScrollViewHorizontalComponent

schema {
  query: Query
}

